packages/beacon-node/src/chain/chain.ts

```diff
@@ -388,40 +389,71 @@ export class BeaconChain implements IBeaconChain {
    * - Some old states (including to-be justified checkpoint) may / must be dropped from the cache
    * - Thus, there is no guarantee that the state for a justified checkpoint will be available in the cache
    */
-  private closestStateToCheckpoint(checkpoint: CheckpointWithHex): CachedBeaconStateAllForks {
+  private closestJustifiedBalancesStateToCheckpoint(checkpoint: CheckpointWithHex): CachedBeaconStateAllForks {
     const state = this.checkpointStateCache.get(checkpoint);
     if (state) {
       // most of the time it should go here
       return state;
     }
 
-    // If it's not found, then find the oldest state in the same chain as this one
-    const checkpointSlot = computeStartSlotAtEpoch(checkpoint.epoch);
-    const head = this.forkChoice.getHead();
+    // Find a state in the same branch of checkpoint at same epoch. Balances should exactly the same
+    for (const descendantBlock of this.forkChoice.forwardIterateDescendants(checkpoint.rootHex)) {
```

Comment:
should we consume `forwardIterateDescendants` only once to improve performance a bit? either extract to an array, or that function could return an array itself

Reply:
Since it's very unlikely to hit this code path I would not do that for simplicity. Metrics will clearly show if we have to run the iterator twice too often, only then optimize